ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/CCodeGenerator.java

```diff
@@ -59,121 +62,131 @@ public CCodeGenerator() {
-                if (ConverterUtil.hasContent(ds3TypeEntry.getElements())) {
-                    final Type typeEntry = TypeConverter.toType(ds3TypeEntry);
-                    generateTypeTemplate(typeEntry, "TypeElement.ftl");
-                    generateTypeTemplate(typeEntry, "FreeTypeElementPrototype.ftl");
-                    generateTypeTemplate(typeEntry, "FreeTypeElement.ftl");
-                }
+                allStructsBuilder.add(StructConverter.toStruct(ds3TypeEntry));
             }
         }
+        return allStructsBuilder.build();
+    }
 
-        // Generate Requests last because they depend on the code generated above
-        if (ConverterUtil.hasContent(spec.getRequests())) {
-            for (final Ds3Request request : spec.getRequests()) {
-                generateRequest(request);
+    private static boolean containsExistingStructs(final Struct struct, final Set<String> existingStructs) {
+        for (final StructMember structMember: struct.getVariables()) {
+            final String testType = StringUtils.stripEnd(structMember.getType(), "*");
+            if (!existingStructs.contains(testType)) {
+                return false;
             }
         }
+        return true;
     }
 
-    public void generateRequest(final Ds3Request ds3Request) throws IOException, TemplateException {
-        Template requestTemplate = null;
-        Request request = null;
-
-        if (ds3Request.getClassification() == Classification.amazons3) {
-            request = RequestConverter.toRequest(ds3Request);
-            requestTemplate = config.getTemplate("AmazonS3InitRequestHandler.ftl");
-        } else if (ds3Request.getClassification() == Classification.spectrads3) {
-            LOG.info("AmazonS3 request");
-            // TODO
-        } else if (ds3Request.getClassification() == Classification.spectrainternal) /* TODO && codeGenType != production */ {
-            LOG.debug("Skipping Spectra internal request");
-            return;
-        } else {
-            throw new TemplateException("Unknown dDs3Request Classification: " + ds3Request.getClassification().toString(), Environment.getCurrentEnvironment());
-        }
+    /**
+     * Return structs which contain only primitive types first, and then cascade for structs that contain other structs
+     *
+     * @throws java.text.ParseException
+     */
+    public static ImmutableList<Struct> getStructsOrderedList(final Ds3ApiSpec spec) throws ParseException {
+        final ImmutableList.Builder<Struct> orderedStructsBuilder = ImmutableList.builder();
+        final Queue<Struct> allStructs = new LinkedList(getAllStructs(spec));
+        final Set<String> existingTypes = new HashSet<>();
+        int skippedStructsCount = 0;
+        while (!allStructs.isEmpty()) {
+            final int allStructsSize = allStructs.size();
+            final Struct structEntry = allStructs.peek();
+            if (orderedStructsBuilder.build().contains(structEntry)) {
```

Comment:
I thought we had talked about changing this to look it up on the set rather than iterating over the list.